---
title: "JoesFlow"
output:
  md_document:
    variant: markdown_github
vignette: >
  %\VignetteIndexEntry{JoesFlow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


The goal of JoesFlow is to analyze high-dimensional single cell data from flow cytometry, scRNA-seq, CITE-seq and any kind of single cell matrix data. JoesFlow utilizes novel scRNA-seq dimension reduction techniques to generate interpretable and informative visualizations that incorporate all aspects of a dataset in an unbiased manner.

```{r libs, echo=FALSE, message=FALSE, results='hide'}
library(DT)
library(RColorBrewer)
library(ggsci)
library(ggplot2)
library(ggrepel)
library(reshape2)
library(gridExtra)
library(uwot)
library(ComplexHeatmap)
library(circlize)

colors_clusters = c(pal_d3("category10")(10), pal_d3("category20b")(20), pal_igv("default")(51))
colors_samples = c(brewer.pal(5, "Set1"), brewer.pal(8, "Dark2"), pal_igv("default")(51))

```

```{r setup, fig.height = 10, fig.width = 12, fig.align = "center"}
library(JoesFlow)
run_app()
```

## JoesFlow Functions

```{r test_data}

sample_data=read.csv("/Users/devlij03/Desktop/JoesFlow/tests/flow_test.csv", sep=',', header=T)
meta_data=read.csv("/Users/devlij03/Desktop/JoesFlow/tests/metadata.csv", header=T, sep=',')

datatable(sample_data[1:100,1:20])

datatable(meta_data)
```

```{r pca_source, echo=FALSE}

cluster_pca=function(data, meta, k) {
  data_mat2=data[,-1]
  ids=data[,1]
  data_mat2=data.matrix(data_mat2)

  pp=prcomp(data_mat2, scale=T)
  pp
  
  kmeaner=kmeans(data_mat2, k)
  kk=paste0("C", kmeaner$cluster)
  
  kmeans=as.character(kk)
  
  #actual PCA
  plotter=data.frame(PC1=pp$x[,1], PC2=pp$x[,2], SampleID=ids)
  plotter$Kmeans=as.character(kmeans)
  
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
    
  PoV <- pp$sdev^2/sum(pp$sdev^2)
    
  colors_samples2=colors_samples
   
  gg1=ggplot(plotter, aes(PC1, PC2, color=Group)) +
    geom_point() + theme_bw() +
    scale_color_manual(values=colors_samples2) +
    xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
    ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
    
  
  #kmeans pca
  gg2=ggplot(sample(plotter), aes(PC1, PC2, color=Kmeans)) +
      geom_point() + theme_bw() +
      scale_color_manual(values=colors_clusters) +
      xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
      ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
      theme(axis.text=element_text(color='black', size=14),
            axis.title=element_text(color='black', size=16))
  
  #sample based pca
    
  totaler=data.frame(table(ids))
  k_df = data.frame(table(kmeans, ids))
  k_mat = dcast(k_df, ids ~ kmeans)
  
  k_mat=k_mat[,-1]
  k_add=apply(k_mat, 2, function(x){(x/totaler$Freq)*100})
  
  pp=prcomp(k_add)
  plotter=data.frame(PC1=pp$x[,1], PC2=pp$x[,2], SampleID=totaler$ids)
      
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
    
  PoV <- pp$sdev^2/sum(pp$sdev^2)
  
  colors_samples2=colors_samples
  
  pp1=ggplot(plotter, aes(PC1, PC2, color=Group, label=SampleID)) +
    geom_point(size=4) + theme_bw() +
    geom_label_repel(size=6) +
    scale_color_manual(values=colors_samples2) +
    xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
    ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
  
  #loadings
  
  pp_load=pp$rotation
  plotter2=data.frame(PC1=pp_load[,1], PC2=pp_load[,2], Label=rownames(pp_load))
  pp2=ggplot(plotter2, aes(PC1, PC2, color=Label, label=Label)) +
    theme_bw() +
    geom_label_repel(size=6) +
    scale_color_manual(values=colors_clusters) +
    #xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
    #ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
  
  return(grid.arrange(gg1,gg2,pp1,pp2, nrow=2))
}

cluster_umap=function(data, meta, k) {
  data_mat2=data[,-1]
  ids=data[,1]
  data_mat2=data.matrix(data_mat2)

  mnist_umap <- umap(data_mat2, pca = 15, fast_sgd = TRUE)
  colnames(mnist_umap)=c("UMAP_1", "UMAP_2")
  
  kmeaner=kmeans(data_mat2, k)
  kk=paste0("C", kmeaner$cluster)
  
  kmeans=as.character(kk)
  
  #actual UMAP
  plotter=data.frame(UMAP_1=mnist_umap[,1], UMAP_2=mnist_umap[,2], SampleID=ids)
  plotter$Kmeans=as.character(kmeans)
  
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
    
  colors_samples2=colors_samples
   
  gg1=ggplot(plotter, aes(UMAP_1, UMAP_2, color=Group)) +
    geom_point() + theme_bw() +
    scale_color_manual(values=colors_samples2) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
    
  
  #kmeans umap
  gg2=ggplot(sample(plotter), aes(UMAP_1, UMAP_2, color=Kmeans)) +
      geom_point() + theme_bw() +
      scale_color_manual(values=colors_clusters) +
      theme(axis.text=element_text(color='black', size=14),
            axis.title=element_text(color='black', size=16))
  
  #sample based pca
    
  totaler=data.frame(table(ids))
  k_df = data.frame(table(kmeans, ids))
  k_mat = dcast(k_df, ids ~ kmeans)
  
  k_mat=k_mat[,-1]
  k_add=apply(k_mat, 2, function(x){(x/totaler$Freq)*100})
  
  pp=prcomp(k_add)
  plotter=data.frame(PC1=pp$x[,1], PC2=pp$x[,2], SampleID=totaler$ids)
      
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
    
  PoV <- pp$sdev^2/sum(pp$sdev^2)
  
  colors_samples2=colors_samples
  
  pp1=ggplot(plotter, aes(PC1, PC2, color=Group, label=SampleID)) +
    geom_point(size=4) + theme_bw() +
    geom_label_repel(size=6) +
    scale_color_manual(values=colors_samples2) +
    xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
    ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
  
  #loadings
  
  pp_load=pp$rotation
  plotter2=data.frame(PC1=pp_load[,1], PC2=pp_load[,2], Label=rownames(pp_load))
  pp2=ggplot(plotter2, aes(PC1, PC2, color=Label, label=Label)) +
    theme_bw() +
    geom_label_repel(size=6) +
    scale_color_manual(values=colors_clusters) +
    #xlab(paste0("PC1 (Explained Variance ", round(PoV[1],4)*100, "%)")) +
    #ylab(paste0("PC2 (Explained Variance ", round(PoV[2],4)*100, "%)")) +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16))
  
  return(grid.arrange(gg1,gg2,pp1,pp2, nrow=2))
}

composition_plot=function(data, meta, k) {
  data_mat2=data[,-1]
  ids=data[,1]
  data_mat2=data.matrix(data_mat2)

  kmeaner=kmeans(data_mat2, k)
  kk=paste0("C", kmeaner$cluster)
  
  kmeans=as.character(kk)
    
  totaler=data.frame(table(ids))
  k_df = data.frame(table(kmeans, ids))
  k_mat = dcast(k_df, ids ~ kmeans)
  
  k_mat=k_mat[,-1]
  k_add=apply(k_mat, 2, function(x){(x/totaler$Freq)*100})
  
  plotter=data.frame(k_add, SampleID=totaler$ids)
  
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
  plotter_melt=melt(data=plotter, id.vars=c("SampleID", "Group"))
      
  k_cols=length(unique(plotter_melt$Group))
  colors_use = c(colors_clusters[1:k_cols], colors_samples)
  
  g1=ggplot(plotter_melt, aes(SampleID, value, fill=variable)) + 
    #geom_tile(aes(x=SampleID,y=105,fill=Group, height=5), show.legend = F) +
    geom_col()+ scale_fill_manual(values=colors_clusters) + theme_bw() +
    guides(fill = guide_legend("Cluster")) + ylab("Cluster Percentage %") +
    theme(axis.text=element_text(color='black', size=14),
          axis.title=element_text(color='black', size=16)) +
    facet_wrap(~Group, ncol=k_cols, scales="free")
  grid.arrange(g1)
  total_col=ncol(plotter)
  datatable(data.frame(plotter[,(total_col-1):total_col], plotter[,1:(total_col-2)]))
}

heat_plot=function(data, meta, k) {
  data_mat2=data[,-1]
  ids=data[,1]
  data_mat2=data.matrix(data_mat2)
  
  kmeaner=kmeans(data_mat2, k)
  kk=paste0("C", kmeaner$cluster)
  
  kmeans=as.character(kk)
  
  plotter=data.frame(SampleID=ids)
  plotter$Kmeans=as.character(kmeans)
  
  plotter$Group=as.character(plotter$SampleID)
  samps=as.character(unique(plotter$SampleID))
  for(jj in 1:length(samps)){
    grouper=subset(meta, ID==samps[jj])
    grouper=as.character(grouper$Group[1])
    
    plotter$Group[plotter$SampleID==samps[jj]]<-grouper
  }
  
  samp_samp=sample(1:nrow(plotter), 500)
  plotter_sub=plotter[samp_samp,]
  data_mat2_sub=data_mat2[samp_samp,]
  
  colorer2=c()
  for(j in 1:length(plotter_sub$Group)){
    colorer2[j]=colors_samples[j]
    names(colorer2)[j]=plotter_sub$Group[j]
    }
  ha=columnAnnotation(Group=plotter_sub$Group, col=list(Group=colorer2))
  
  Heatmap(scale(t(data_mat2_sub)), show_row_names = T, show_column_names = F,
              top_annotation = ha,
              heatmap_legend_param = list(title = "Scaled Value"),
              cluster_rows = T, cluster_columns = T, row_names_side = 'left',
              column_names_gp = gpar(fontsize=7),
              row_names_gp = gpar(fontsize=10),
              row_title_gp = gpar(fontsize = 10),
              row_names_max_width = unit(10,'cm'),
              use_raster = T,
              column_split=plotter_sub$Kmeans,
              cluster_column_slices=F,
              #column_split = splitter,
              #left_annotation = ha,
              col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")))
}

```

```{r pca, fig.height = 10, fig.width = 12, fig.align = "center"}

cluster_pca(sample_data, meta_data, 10)

```

```{r umap, fig.height = 10, fig.width = 12, fig.align = "center"}

cluster_umap(sample_data, meta_data, 10)

```

```{r comp_plot, fig.height = 6, fig.width = 8, fig.align = "center"}

composition_plot(sample_data, meta_data, 15)

```

```{r heat_plot, fig.height = 10, fig.width = 12, fig.align = "center"}

heat_plot(sample_data, meta_data, 10)

```